### PR TITLE
Tony/17_save_json

### LINKED ISSUE #17: MD WIDGETS: Save user input data to a local JSON file
## Description

Add an additional parameter to the Submit button and let users save the user input data as a local JSON file. Make this submission method the default, while also providing options for users to submit via POST to a URL or POD.

### PR DESCRIPTION
**Pull request by tonypioneer**
# Pull Request Details

## What issue does this PR address

- Modified the button widget and allowed users to save survey results to a local JSON file or submit them to a server.

- Please note that if you would like to test save-to-server function, please run `make server` first. Surveys 1 and 2 provide the demos of saving survey results to a local JSON file, and Survey 3 provides the demo of submitting the results to a server.

- Link to associated issue: #17

## Checklist

Complete the check-list below to ensure your branch is ready for PR.

Flutter Style Guide: https://survivor.togaware.com/gnulinux/flutter-style.html

- [ ] Screenshots included in linked issue
- [x] Changes adhere to the team style and coding guideline
- [x] No confidential information
- [x] No duplicated content
- [x] No lint check errors related to your changes (`make prep` or `flutter analyze lib`)
- [ ] Pre-exisiting lint errors noted: [HERE]
- [x] Tested on at least one device
  - [ ] Android Phone
  - [ ] Android Emulator
  - [ ] Chrome on Android
  - [x] Chrome
  - [ ] iOS
  - [ ] Linux
  - [x] MacOS
  - [ ] Windows
- [x] Added 2 reviewers (or 1 for private repositories then they add another)

## Finalising

Once PR discussion is complete and 2 reviewers have approved:

- [ ] Merge dev into the branch
- [ ] Resolve any conflicts
- [ ] Add one line summary into CHANGELOG.md
- [ ] Bump appropriate version number in pubspec.yaml
- [ ] Push to git repository and review
- [ ] Merge PR into dev

### CODE REVIEW

File: lib/src/constants/pkg.dart [hunk 1]
```diff
@@ -38,6 +38,9 @@ const String assetsPath = 'assets';
 const String mediaPath = 'assets/media';
 const String surveyAsset = 'assets/surveys.md';
 
+const String defaultFileName = 'result.json';
```
Comment (gjwgit) [inline lib/src/constants/pkg.dart]: Perhaps call it `survey.json`

File: lib/src/widgets/button_widget.dart [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+///
+/// Authors: Tony Chen
+
+import 'dart:async';
```
Comment (gjwgit) [inline lib/src/widgets/button_widget.dart:31]: Please see https://survivor.togaware.com/gnulinux/flutter-style-imports.html for our style for grouping imports

File: lib/src/widgets/button_widget.dart [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+        actionParameter = defaultFileName;
+      }
+    } else {
+      buttonText = defaultButtonText;
+      actionType = 0;
+      actionParameter = defaultFileName;
+    }
+  }
+
+  /// Collects the user responses from the state map.
```
Comment (gjwgit) [inline lib/src/widgets/button_widget.dart:95]: Our style encourages a blank line after comments: https://survivor.togaware.com/gnulinux/flutter-style-comments.html

File: lib/src/widgets/button_widget.dart [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+  Map<String, dynamic> _collectData() {
+    // Access the state variables and collect data
```
Comment (gjwgit) [inline lib/src/widgets/button_widget.dart]: Minor thing but we should end sentences with a full stop https://survivor.togaware.com/gnulinux/flutter-style-comments.html

File: lib/src/widgets/button_widget.dart [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+    _inputValues.forEach((key, value) {
+      responses[key] = value;
+    });
+
+    return responses;
+  }
+
+  Future<void> _handleButtonPress() async {
+    final data = _collectData();
+
```
Comment (gjwgit) [inline lib/src/widgets/button_widget.dart:160]: OPTIONAL: The following is quite long and nested. Is there an opportunity to define some local functions and use them here so that we can read the code more readily.

**Review by gjwgit — COMMENTED** [review]
Suggest removing all of the files that will be created by `flutter create .` from the example folder, and only keep those that you specifically need to add/modify.

**Review by gjwgit — COMMENTED** [review]
I prefer on Linux that I am provided a single file chooser where I can choose both the location and file name in the one popup. Currently I get a popup to ask for the file name:

image

And then a folder chooser

image

In rattle the filename and folder are done in a single popup:

image

**Comment by gjwgit** [comment]
Could you also pretty print the output JSON when it is saved to file. Thanks

File: lib/src/constants/pkg.dart (hunk 1 above)
Comment (gjwgit) [reply lib/src/constants/pkg.dart]: Actually, it is not always a survey! Is it possible to default the filename to be the menu item name. In the example this will be `survey_1.json` or `survey_2.json` or `survey_3.json` depending which survey. The name is derived fromt he menu item/survey name and normalised for a nice filename (lowercase and replace space with underscore).

Comment (tonypioneer) [reply lib/src/constants/pkg.dart]: > Actually, it is not always a survey! Is it possible to default the filename to be the menu item name. In the example this will be `survey_1.json` or `survey_2.json` or `survey_3.json` depending which survey. The name is derived from the menu item/survey name and normalised for a nice filename (lowercase and replace space with underscore).

Yes, it is a very good idea. I'll implement it.

**Comment by tonypioneer** [comment]
> I prefer on Linux that I am provided a single file chooser where I can choose both the location and file name in the one popup. Currently I get a popup to ask for the file name:
> 
> image
> 
> And then a folder chooser
> 
> image
> 
> In rattle the filename and folder are done in a single popup:
> 
> image



> I prefer on Linux that I am provided a single file chooser where I can choose both the location and file name in the one popup. Currently I get a popup to ask for the file name:
> 
> image
> 
> And then a folder chooser
> 
> image
> 
> In rattle the filename and folder are done in a single popup:
> 
> image

This function has been implemented.
image

**Review by gjwgit — APPROVED** [review]
Let's get this merged and resolve any other issues separately.

**Review by gjwgit — APPROVED** [review]
Decided to sneak in the pretty print. Please review.

File: lib/src/widgets/audio_widget.dart
Comment (gjwgit) [inline lib/src/widgets/audio_widget.dart:1]: Best to keep unrelated changes in a separate PR so we can more easily review just the key functionality for this PR.

**Review by anushkavidanage — APPROVED** [review]
Works great. Thanks Tony
